transport/transporters/s3/transporter/transporter.go

```diff
@@ -0,0 +1,351 @@
+		log,
+		client,
+		bucketName,
+		keySpace,
+		retryPolicy,
+	}
+}
+
+// NewTransporter returns a s3 transporter
+func NewTransporter(shutdownHandler shutdown.ShutdownHandler,
+	inputChan <-chan transport.Batch,
+	txnsWritten chan<- *ordered_map.OrderedMap,
+	statsChan chan stats.Stat,
+	log logrus.Entry,
+	id int,
+	bucketName string,
+	keySpace string,
+	retryPolicy backoff.BackOff,
+	awsRegion *string,
+	awsAccessKeyId *string,
+	awsSecretAccessKey *string,
+	endpoint *string,) transport.Transporter {
+
+	awsConfig := &aws.Config{
+		Region:      		aws.String(*awsRegion),
+		S3ForcePathStyle: 	aws.Bool(true),
+		Endpoint:    		aws.String(*endpoint),
+		MaxRetries:		    aws.Int(0), // We disable the client retry policy because of our own retry logic
+	}
+
+	if *awsAccessKeyId != "" || *awsSecretAccessKey != "" {
+		// Force static credentials from pg-bifrost configuration.
+		// Note: if we expect to fail and not infer credentials if only one of the ID or Key was specified
+		awsConfig.Credentials = credentials.NewStaticCredentials(*awsAccessKeyId, *awsSecretAccessKey, "")
+	}
+
+	if *endpoint != "" {
+		// If specifying a custom endpoint (such as for localstack) then configure that and use Path Style
+		awsConfig.Endpoint = aws.String(*endpoint)
+		awsConfig.S3ForcePathStyle = aws.Bool(true)
+	}
+
+	sess := session.Must(session.NewSession(awsConfig))
+	client := s3.New(sess)
+
+	return NewTransporterWithInterface(
+		shutdownHandler,
+		inputChan,
+		txnsWritten,
+		statsChan,
+		log,
+		id,
+		bucketName,
+		keySpace,
+		client,
+		retryPolicy)
+}
+
+// shutdown idempotently closes the output channel
+func (t *S3Transporter) shutdown() {
+	t.log.Info("shutting down transporter")
+	t.shutdownHandler.CancelFunc() // initiate shutdown on other modules as well
+
+	if r := recover(); r != nil {
+		t.log.Warnf("Recovered in S3Transporter %s", r)
+	}
+
+	defer func() {
+		// recover if channel is already closed
+		recover()
+	}()
+
+	t.log.Debug("closing progress channel")
+	close(t.txnsWritten)
+}
+
+
+// transportWithRetry does a PUT on a full batch as a single key/file to S3
+func (t *S3Transporter) transportWithRetry(ctx context.Context, messagesSlice []*marshaller.MarshalledMessage) (error, bool) {
+	var cancelled bool
+	var buf bytes.Buffer
+	gz := gzip.NewWriter(&buf)
+
+	select {
+	case <-ctx.Done():
+		t.log.Debug("received terminateCtx cancellation")
+		cancelled = true
+	default:
+	}
+
+	// add all messages into a gzipped buffer
+	for _, msg := range messagesSlice {
+		if _, err := gz.Write(msg.Json); err != nil {
+			return err, cancelled
+		}
+		gz.Write([]byte("\n"))
+	}
+
+	if err := gz.Flush(); err != nil {
+		return err, cancelled
+	}
+	if err := gz.Close(); err != nil {
+		return err, cancelled
+	}
+
+	gz.Reset(ioutil.Discard)
+
+	firstWalStart := messagesSlice[0].WalStart
+
+	// Free up messagesSlice now that we're done with it
+	messagesSlice = nil
+
+	// Clear out every allocation in the conversion
+	byteArray := buf.Bytes()
```

Comment:
Does it make sense to add expected perf logs to guard against introducing memory leaks?

Reply:
In this scenario, there wouldn't be a leak, as these variables are overwritten on every subsequent batch. The memory impact scales linearly and with an upper-bound, whether or not we make these optimizations.